### PR TITLE
feat: Add Charmed Kubeflow Terraform solution

### LINKED ISSUE #1073: Create a terraform solution for Kubeflow
### Context

We need to provide a terraform solution for deploying Kubeflow that will expose all required configurations.

### What needs to get done

1. Create a repo for analytics terraform solution
2. Create a solution for Kubeflow

### Definition of Done

Kubeflow can be deployed using a Terraform solution.

### PR DESCRIPTION
**Pull request by orfeas-k**
This PR adds a Terraform solution module for Charmed Kubeflow bundle. Please refer to the README.md file for instructions on how this works.

### COS integration
Here's the [integrate CKF with COS guide](https://charmed-kubeflow.io/docs/integrate-with-cos)  for reference. 

### Outputs
I defined the outputs of the deployed mode, plus the grafana-agent-k8s app when cos configuration is enabled since [CC006 spec](https://docs.google.com/document/d/1EG71A2pJ244PQRaGVzGj7Mx2B_bzE4U_OSqx4eeVI1E/edit) mentions:
>Solution modules shall output the names of deployed applications that can be integrated with from outside the solution. The modules shall also output the names of all integration endpoints that can be integrated with from outside the solution.

This information should be enough for someone wanting to integrate with COS.

### Testing intrucstions
Here are some instructions for reference, but again, please refer to the README.md for more details, since this will serve as instructions for any user.
```
# Set an environment with juju 3.4 and Microk8s 1.29
# Install terraform
$ cd kubeflow
$ terraform init
$ terraform apply -var cos_configuration=true --auto-approve
```

Closes canonical/bundle-kubeflow/issues/1073

### CODE REVIEW

**Review by misohu — COMMENTED** [review]
I have checked the PR and tried to deploy it with different setting. For solo kubeflow without cos I had no problems. For cos the apply command succeeds but the apps are in following state. 

```
Model     Controller  Cloud/Region      Version  SLA          Timestamp
kubeflow  uk8sx       my-k8s/localhost  3.4.6    unsupported  11:15:11Z

App                         Version                  Status       Scale  Charm                    Channel          Rev  Address         Exposed  Message
admission-webhook                                    active           1  admission-webhook        1.9/stable       344  10.152.183.91   no       
argo-controller                                      active           1  argo-controller          3.4/stable       545  10.152.183.33   no       
dex-auth                                             active           1  dex-auth                 2.39/stable      548  10.152.183.202  no       
envoy                                                active           1  envoy                    2.2/stable       263  10.152.183.92   no       
grafana-agent-k8s-kubeflow  0.32.1                   waiting          1  grafana-agent-k8s        latest/stable     45  10.152.183.183  no       installing agent
istio-ingressgateway                                 active           1  istio-gateway            1.22/stable     1280  10.152.183.50   no       
istio-pilot                                          active           1  istio-pilot              1.22/stable     1169  10.152.183.185  no       
jupyter-controller                                   active           1  jupyter-controller       1.9/stable      1038  10.152.183.39   no       
jupyter-ui                                           active           1  jupyter-ui               1.9/stable       961  10.152.183.191  no       
katib-controller                                     waiting          1  katib-controller         0.17/stable      750  10.152.183.246  no       installing agent
katib-db                                             waiting          1  mysql-k8s                8.0/stable       180  10.152.183.93   no       installing agent
katib-db-manager                                     waiting          1  katib-db-manager         0.17/stable      713  10.152.183.17   no       installing agent
katib-ui                                             active           1  katib-ui                 0.17/stable      713  10.152.183.243  no       
kfp-api                                              maintenance      1  kfp-api                  2.2/stable      1552  10.152.183.250  no       Creating K8S resources
kfp-db                      8.0.37-0ubuntu0.22.04.3  active           1  mysql-k8s                8.0/stable       180  10.152.183.109  no       
kfp-metadata-writer                                  active           1  kfp-metadata-writer      2.2/stable       617  10.152.183.41   no       
kfp-persistence                                      active           1  kfp-persistence          2.2/stable      1560  10.152.183.134  no       
kfp-profile-controller                               active           1  kfp-profile-controller   2.2/stable      1518  10.152.183.49   no       
kfp-schedwf                                          active           1  kfp-schedwf              2.2/stable      1571  10.152.183.40   no       
kfp-ui                                               active           1  kfp-ui                   2.2/stable      1555  10.152.183.37   no       
kfp-viewer                                           active           1  kfp-viewer               2.2/stable      1586  10.152.183.71   no       
kfp-viz                                              active           1  kfp-viz                  2.2/stable      1504  10.152.183.212  no       
knative-eventing                                     active           1  knative-eventing         1.12/stable      459  10.152.183.247  no       
knative-operator                                     active           1  knative-operator         1.12/stable      433  10.152.183.221  no       
knative-serving                                      active           1  knative-serving          1.12/stable      487  10.152.183.229  no       
kserve-controller                                    active           1  kserve-controller        0.13/stable      646  10.152.183.228  no       
kubeflow-dashboard                                   active           1  kubeflow-dashboard       1.9/stable       659  10.152.183.154  no       
kubeflow-profiles                                    active           1  kubeflow-profiles        1.9/stable       458  10.152.183.205  no       
kubeflow-roles                                       active           1  kubeflow-roles           1.9/stable       240  10.152.183.158  no       
kubeflow-volumes                                     active           1  kubeflow-volumes         1.9/stable       348  10.152.183.30   no       
metacontroller-operator                              active           1  metacontroller-operator  3.0/stable       311  10.152.183.163  no       
minio                       res:oci-image@5102166    active           1  minio                    ckf-1.9/stable   347  10.152.183.237  no       
mlmd                                                 active           1  mlmd                     ckf-1.9/stable   213  10.152.183.108  no       
oidc-gatekeeper                                      active           1  oidc-gatekeeper          ckf-1.9/stable   423  10.152.183.46   no       
pvcviewer-operator                                   active           1  pvcviewer-operator       1.9/stable       157  10.152.183.103  no       
tensorboard-controller                               active           1  tensorboard-controller   1.9/stable       355  10.152.183.189  no       
tensorboards-web-app                                 active           1  tensorboards-web-app     1.9/stable       343  10.152.183.173  no       
training-operator                                    active           1  training-operator        1.8/stable       503  10.152.183.148  no       

Unit                           Workload     Agent  Address     Ports          Message
admission-webhook/0*           active       idle   10.1.41.16                 
argo-controller/0*             active       idle   10.1.41.14                 
dex-auth/0*                    active       idle   10.1.41.51                 
envoy/0*                       active       idle   10.1.41.34                 
grafana-agent-k8s-kubeflow/0*  blocked      idle   10.1.41.22                 grafana-dashboards-provider: off, grafana-cloud-config: off
istio-ingressgateway/0*        active       idle   10.1.41.27                 
istio-pilot/0*                 active       idle   10.1.41.41                 
jupyter-controller/0*          active       idle   10.1.41.29                 
jupyter-ui/0*                  active       idle   10.1.41.28                 
katib-controller/0*            maintenance  idle   10.1.41.38                 Reconciling charm: executing component katib-controller-pebble-service
katib-db-manager/0*            waiting      idle   10.1.41.35                 Waiting for relational-db data
katib-db/0*                    error        idle   10.1.41.46                 hook failed: "metrics-endpoint-relation-created"
katib-ui/0*                    active       idle   10.1.41.40                 
kfp-api/0*                     maintenance  idle   10.1.41.18                 Creating K8S resources
kfp-db/0*                      active       idle   10.1.41.12                 Primary
kfp-metadata-writer/0*         active       idle   10.1.41.17                 
kfp-persistence/0*             active       idle   10.1.41.26                 
kfp-profile-controller/0*      active       idle   10.1.41.23                 
kfp-schedwf/0*                 active       idle   10.1.41.13                 
kfp-ui/0*                      active       idle   10.1.41.36                 
kfp-viewer/0*                  active       idle   10.1.41.55                 
kfp-viz/0*                     active       idle   10.1.41.53                 
knative-eventing/0*            active       idle   10.1.41.43                 
knative-operator/0*            active       idle   10.1.41.48                 
knative-serving/0*             active       idle   10.1.41.44                 
kserve-controller/0*           active       idle   10.1.41.31                 
kubeflow-dashboard/0*          active       idle   10.1.41.37                 
kubeflow-profiles/0*           active       idle   10.1.41.15                 
kubeflow-roles/0*              active       idle   10.1.41.33                 
kubeflow-volumes/0*            active       idle   10.1.41.24                 
metacontroller-operator/0*     active       idle   10.1.41.30                 
minio/1*                       active       idle   10.1.41.20  9000-9001/TCP  
mlmd/0*                        active       idle   10.1.41.25                 
oidc-gatekeeper/0*             active       idle   10.1.41.5                  
pvcviewer-operator/0*          active       idle   10.1.41.11                 
tensorboard-controller/0*      active       idle   10.1.41.39                 
tensorboards-web-app/0*        active       idle   10.1.41.42                 
training-operator/0*           active       idle   10.1.41.49                 
```

I think its expected as I am experiencing the same in my mlflow PR. Let me know what you think @orfeas-k

File: kubeflow/applications.tf [hunk 1]
```diff
@@ -0,0 +1,243 @@
+
+module "admission_webhook" {
+  source     = "git::https://github.com/canonical/admission-webhook-operator//terraform?ref=track/1.9"
+  model_name = var.create_model ? juju_model.kubeflow[0].name : local.model_name
```
Comment (ghislainbourgeois) [inline kubeflow/applications.tf:4]: `model_name` has been changed to `model` in the spec, as newer version of the provider will slowly change towards using UUIDs. This comment is valid for the rest of the project and the underlying modules.

File: kubeflow/relations.tf [hunk 2]
```diff
@@ -0,0 +1,490 @@
+
+resource "juju_integration" "argo_controller_minio" {
+  model = var.create_model ? juju_model.kubeflow[0].name : local.model_name
+
+  application {
+    name     = module.argo_controller.app_name
+    endpoint = "object-storage"
```
Comment (ghislainbourgeois) [inline kubeflow/relations.tf]: In general, the endpoint should be provided by the module:
```suggestion
    endpoint = module.argo_controller.requires.object-storage
```

This of course requires an update to the module.

**Review by ghislainbourgeois — COMMENTED** [review]
General comments related to spec CC006. This should not prevent you from merging as-is and revisiting in the future when the spec is finalized.

File: kubeflow/relations.tf (hunk 2 above)
Comment (orfeas-k) [reply kubeflow/relations.tf]: The modules already support this so let me update it.

File: kubeflow/applications.tf (hunk 1 above)
Comment (orfeas-k) [reply kubeflow/applications.tf:4]: Thanks for catching this. I 'll update this solution module to use `model` and then file an issue to update all charm modules at some point with the proper attribute name.

**Comment by orfeas-k** [comment]
@misohu that issue is due to https://github.com/canonical/mysql-k8s-operator/issues/504. This has been fixed, but my guess it is not yet in `stable`.

File: kubeflow/applications.tf (hunk 1 above)
Comment (orfeas-k) [reply kubeflow/applications.tf:4]: Filed issue https://github.com/canonical/analytics-terraform-solutions/issues/8

File: kubeflow/relations.tf (hunk 2 above)
Comment (orfeas-k) [reply kubeflow/relations.tf]: updated

**Comment by orfeas-k** [comment]
Closing in favor of canonical/charmed-kubeflow-solutions#2

**Pull request closed without merging**